packages/theme-cart/__tests__/theme-cart.test.js

```diff
@@ -187,6 +203,62 @@ describe('addItem()', () => {
   });
 });
 
+describe('addItemFromForm()', () => {
+  beforeEach(() => {
+    document.body.innerHTML = `
+      <form id="valid-form">
+        <select name="id">
+          <option selected="selected" value="32497200136297">small</option>
+          <option disabled="disabled" value="32497200169065">large</option>
+        </select>
+        <input type="number" name="quantity" value="3" />
+        <input type="text" name="properties[Message]" value="derp" />
+      </form>
+      <form id="empty-form"></form>`;
+
+    fetchMock.mock('/cart/add.js', mockCartAddFromForm);
+  });
+
+  afterEach(fetchMock.restore);
+
+  test('throws an error if parameter is not a valid form', () => {
+    const validForm = document.getElementById('valid-form');
+
+    expect(() => cart.addItemFromForm(validForm)).not.toThrowError();
+    expect(() => cart.addItemFromForm().toThrowError(TypeError));
+    expect(() => cart.addItemFromForm({}).toThrowError(TypeError));
+    expect(() => cart.addItemFromForm('123456')).toThrowError(TypeError);
+  });
+
+  test('throws an error if form is missing id or quantity', () => {
+    const emptyForm = document.getElementById('empty-form');
+    expect(() => cart.addItemFromForm(emptyForm)).toThrowError(Error);
+  });
+
+  test('throws an error if either id or quantity is invalid', () => {
+    const validForm = document.getElementById('valid-form');
+
+    const invalidForm = document.createElement('form');
+    invalidForm.innerHTML = validForm.innerHTML.replace('32497200136297', 'abcde');
+    expect(() => cart.addItemFromForm(invalidForm)).toThrowError(TypeError);
+
+    invalidForm.innerHTML = validForm.innerHTML.replace('name="quantity" value="3"', 'name="quantity" value="de"');
```

Comment:
I don't know the right answer to this, so I'm going to ask: 
Does it make more sense to create an invalid form instead of replacing values?

Reply:
I created an invalid form now. I didn't before because I wanted one with invalid `id` (but valid `quantity`) and another one with invalid `quantity` (but valid `id`) :)